File: server/src/nodes/dtos/node.dto.ts
import { Type } from 'class-transformer';
import {
  IsDefined,
  IsMongoId,
  IsString,
  Matches,
  ValidateNested,
} from 'class-validator';
import { NodeInfoDto } from './commons';

export class NodeDto {
  @IsMongoId()
  id: string;

  @IsMongoId()
  rootId: string;

  @IsMongoId()
  parent: string;

  @IsString()
  @Matches(/^[a-zA-Z0-9-_]+$/)
  name: string;

  @IsString()
  displayName: string;

  @Type(() => NodeInfoDto)
  @ValidateNested()
  @IsDefined()
  nodeInfo: NodeInfoDto;
}

File: clients/flutter_client/lib/src/http/dtos/common/node_dto.dart
import 'package:flutter_client/src/http/dtos/common/node_info_dto.dart';

class NodeDto {
  final String id;
  final String rootId;
  final String parentId;
  final String name;
  final String displayName;
  final NodeInfoDto nodeInfo;
  
  const NodeDto({
    required this.id,
    required this.rootId,
    required this.parentId,
    required this.name,
    required this.displayName,
    required this.nodeInfo,
  });

  factory NodeDto.fromJson(Map<String, dynamic> json) {
    return NodeDto(
        id: json['id'],
        rootId: json['rootId'],
        parentId: json['parentId'],
        name: json['name'],
        displayName: json['displayName'],
        nodeInfo: NodeInfoDto.fromJson(json['nodeInfo']));
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'rootId': rootId,
      'parentId': parentId,
      'name': name,
      'displayName': displayName,
      'nodeInfo': nodeInfo.toJson(),
    };
  }
}
